File: pseudo_secure_channel/src/network.rs
use std::net::{TcpListener, TcpStream};
use std::io::{Read, Write};

use crate::model::{TransferData, EncryptedPacket};
use crate::crypto::{encrypt, decrypt};
use crate::constants::NETWORK_PORT;

// Simulates the receiving peer
pub fn run_receiver() {
    let address = format!("127.0.0.1:{}", NETWORK_PORT);
    let listener = match TcpListener::bind(&address) {
        Ok(l) => l,
        Err(e) => {
            eprintln!("[Receiver] Failed to bind to {}: {}", address, e);
            return;
        }
    };
    println!("[Receiver] Listening on {}...", address);

    match listener.accept() {
        Ok((mut stream, addr)) => {
            println!("[Receiver] Accepted connection from: {}", addr);

            let mut buffer = Vec::new();
            if let Err(e) = stream.read_to_end(&mut buffer) {
                eprintln!("[Receiver] Failed to read from stream: {}", e);
                return;
            }

            if buffer.is_empty() {
                println!("[Receiver] Received empty data.");
                return;
            }

            // Deserialize the wrapper packet
            let packet: EncryptedPacket = match serde_json::from_slice(&buffer) {
                Ok(p) => p,
                Err(e) => {
                    eprintln!("[Receiver] Failed to deserialize packet: {}", e);
                    // Also log raw bytes for debugging if needed
                    // eprintln!("[Receiver] Raw buffer: {:?}", buffer);
                    return;
                }
            };
            println!("[Receiver] Received packet with nonce: ({} bytes)", packet.nonce.len());

            // Decrypt the payload using the received nonce (and the known PSS)
            match decrypt(&packet.encrypted_payload, &packet.nonce) {
                Ok(decrypted_payload) => {
                    // Deserialize the original TransferData
                    match serde_json::from_slice::<TransferData>(&decrypted_payload) {
                        Ok(data) => {
                            println!("[Receiver] Successfully decrypted data:");
                            println!("  Transaction ID: {}", data.transaction_id);
                            println!("  Created At: {}", data.created_at); // Still present in TransferData
                            println!("  Payload Length: {}", data.payload.len());
                            if let Ok(payload_str) = String::from_utf8(data.payload) {
                                println!("  Payload Preview: {:?}...", payload_str.chars().take(50).collect::<String>());
                            } else {
                                println!("  Payload: (non-UTF8 data)");
                            }
                        },
                        Err(e) => {
                            eprintln!("[Receiver] Failed to deserialize decrypted data: {}", e);
                        }
                   }
                },
                Err(e) => {
                    eprintln!("[Receiver] Failed to decrypt payload: {}", e);
                }
            }
        }
        Err(e) => {
            eprintln!("[Receiver] Failed to accept connection: {}", e);
        }
    }
}

// Simulates the sending peer
pub fn run_sender(data_to_send: TransferData) {
    let address = format!("127.0.0.1:{}", NETWORK_PORT);
    println!("[Sender] Attempting to connect to {}...", address);

    match TcpStream::connect(&address) {
        Ok(mut stream) => {
            println!("[Sender] Connected to receiver.");

            println!("[Sender] Original data:");
            println!("  Transaction ID: {}", data_to_send.transaction_id);
            println!("  Payload Length: {}", data_to_send.payload.len());

            // Serialize the original data
            let serialized_data = match serde_json::to_vec(&data_to_send) {
                Ok(s) => s,
                Err(e) => {
                    eprintln!("[Sender] Failed to serialize data: {}", e);
                    return;
                }
            };

            // Encrypt the serialized data - returns payload AND nonce
            let (encrypted_payload, nonce) = encrypt(&serialized_data);
            println!("[Sender] Encrypted payload length: {}", encrypted_payload.len());
            println!("[Sender] Generated nonce: ({} bytes)", nonce.len());

            // Create the wrapper packet with the nonce
            let packet = EncryptedPacket {
                nonce, // Use the generated nonce
                encrypted_payload,
            };

            // Serialize the wrapper packet
            let serialized_packet = match serde_json::to_vec(&packet) {
                Ok(sp) => sp,
                Err(e) => {
                    eprintln!("[Sender] Failed to serialize packet: {}", e);
                    return;
                }
            };

            // Send the encrypted packet
            if let Err(e) = stream.write_all(&serialized_packet) {
                eprintln!("[Sender] Failed to write to stream: {}", e);
            } else {
                println!("[Sender] Sent encrypted packet ({} bytes).", serialized_packet.len());
            }
            // Shutdown gracefully
             let _ = stream.shutdown(std::net::Shutdown::Write);
        }
        Err(e) => {
            eprintln!("[Sender] Failed to connect: {}", e);
        }
    }
}

File: secure_channel/src/network.py
import socket
import json
from datetime import datetime
import struct

from constants import NETWORK_PORT
from crypto import encrypt, decrypt
from model import EncryptedPacket, TransferData


class DateTimeEncoder(json.JSONEncoder):
    def default(self, obj):
        if isinstance(obj, datetime):
            return obj.isoformat()
        elif isinstance(obj, bytes):
            return list(obj)  # Convert bytes to list for JSON serialization
        return super().default(obj)


def run_receiver():
    address = f"127.0.0.1:{NETWORK_PORT}"
    try:
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.bind(("127.0.0.1", NETWORK_PORT))
        sock.listen(1)
        print(f"[Receiver] Listening on {address}...")

        conn, addr = sock.accept()
        print(f"[Receiver] Accepted connection from: {addr}")

        buffer = b""
        while True:
            data = conn.recv(4096)
            if not data:
                break
            buffer += data

        if not buffer:
            print("[Receiver] Received empty data.")
            return

        # Deserialize the received packet
        packet_data = json.loads(buffer.decode())
        packet = EncryptedPacket(
            nonce=bytes(packet_data["nonce"]),
            encrypted_payload=bytes(packet_data["encrypted_payload"])
        )
        print(f"[Receiver] Received packet with nonce: ({len(packet.nonce)} bytes)")

        try:
            # Decrypt the payload
            decrypted_payload = decrypt(packet.encrypted_payload, packet.nonce)
            
            # Deserialize the decrypted data
            data_json = json.loads(decrypted_payload.decode())
            data = TransferData(
                transaction_id=data_json["transaction_id"],
                payload=bytes(data_json["payload"]),
                created_at=datetime.fromisoformat(data_json["created_at"])
            )
            
            print("[Receiver] Successfully decrypted data:")
            print(f"  Transaction ID: {data.transaction_id}")
            print(f"  Created At: {data.created_at}")
            print(f"  Payload Length: {len(data.payload)}")
            
            try:
                payload_str = data.payload.decode('utf-8')
                print(f"  Payload Preview: {payload_str[:50]}...")
            except UnicodeDecodeError:
                print("  Payload: (non-UTF8 data)")
        
        except Exception as e:
            print(f"[Receiver] Failed to decrypt payload: {e}")
    
    except Exception as e:
        print(f"[Receiver] Error: {e}")
    finally:
        try:
            sock.close()
        except:
            pass


def run_sender(data_to_send: TransferData):
    address = f"127.0.0.1:{NETWORK_PORT}"
    print(f"[Sender] Attempting to connect to {address}...")
    
    try:
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.connect(("127.0.0.1", NETWORK_PORT))
        print("[Sender] Connected to receiver.")
        
        print("[Sender] Original data:")
        print(f"  Transaction ID: {data_to_send.transaction_id}")
        print(f"  Payload Length: {len(data_to_send.payload)}")
        
        # Serialize the data
        serialized_data = json.dumps({
            "transaction_id": data_to_send.transaction_id,
            "payload": list(data_to_send.payload),
            "created_at": data_to_send.created_at.isoformat()
        }).encode()
        
        # Encrypt the data
        encrypted_payload, nonce = encrypt(serialized_data)
        print(f"[Sender] Encrypted payload length: {len(encrypted_payload)}")
        print(f"[Sender] Generated nonce: ({len(nonce)} bytes)")
        
        # Create and serialize the packet
        packet = {
            "nonce": list(nonce),
            "encrypted_payload": list(encrypted_payload)
        }
        
        serialized_packet = json.dumps(packet).encode()
        
        # Send the packet
        sock.sendall(serialized_packet)
        print(f"[Sender] Sent encrypted packet ({len(serialized_packet)} bytes).")
        
        sock.close()
    
    except Exception as e:
        print(f"[Sender] Error: {e}") 